### PR TITLE
Update unprescribe command functionality

### PR DESCRIPTION
**Pull request by szelingtan**
do not merge yet

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java [hunk 1]
```diff
@@ -11,28 +15,42 @@
...
-    public static final String MESSAGE_EMPTY_MED_LIST = "patient: %1$s currently has no "
-            + "prescribed medication";
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": removes all medication from the"
-            + " patient specified "
+    public static final String MESSAGE_REMOVE_MED_SUCCESS = "Removed medication %1$s from patient: %2$s";
+    public static final String MESSAGE_REMOVE_ALL_MED_SUCCESS = "Removed all medication from patient: %1$s";
+    public static final String MESSAGE_EMPTY_MED_LIST = "Patient: %1$s currently has no prescribed medication";
+    public static final String MESSAGE_MED_NOT_FOUND = "Medication %1$s is not found in patient: "
+            + "%2$s's prescription list";
+
```
Comment (Bryce-3D) [inline src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:32]: Hi! Is there any reason for the extra new line here between the `public static final String` s?

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java [hunk 2]
```diff
@@ -44,30 +62,93 @@ public CommandResult execute(Model model) throws CommandException {
...
-                patientToEdit.getAddress(),
-                patientToEdit.getLastVisit(),
-                patientToEdit.getTags(),
-                newMedicines);
+        Set<Medicine> currentMedicines = patientToEdit.getMedicines();
+
+        // Check if medicine list is already empty
+        if (currentMedicines.isEmpty()) {
+            throw new CommandException(String.format(MESSAGE_EMPTY_MED_LIST, Messages.format(patientToEdit)));
+        }
```
Comment (Bryce-3D) [inline src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:70]: Hi! Should the command also throw an exception when the command `unprescribe n all` is run? Since it technically can still run on an empty medicine list and not change the state at all? Is the reason to since it's likely not the user intention or some other reason?

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java [hunk 3]
```diff
@@ -11,28 +15,42 @@
...
+    public static final String MESSAGE_EMPTY_MED_LIST = "Patient: %1$s currently has no prescribed medication";
+    public static final String MESSAGE_MED_NOT_FOUND = "Medication %1$s is not found in patient: "
+            + "%2$s's prescription list";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": removes a medication from the patient specified "
             + "by the index number used in the last patient listing. "
             + "Parameters: INDEX (must be a positive integer) "
-            + "Example: " + COMMAND_WORD + " 1 ";
+            + PREFIX_MEDICINE + "[medicine name] OR "
+            + PREFIX_MEDICINE + "all\n"
```
Comment (Bryce-3D) [inline src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:37]: Hi! What should be the behavior if we want to remove a medicine named "all"? Would it be fine to not have this specific functionality?

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java [hunk 4]
```diff
@@ -44,30 +62,93 @@ public CommandResult execute(Model model) throws CommandException {
...
+            throw new CommandException(String.format(MESSAGE_MED_NOT_FOUND,
+                    medicineToRemove.getMedicineName(), Messages.format(patientToEdit)));
+        }
+
+        Patient editedPatient = createEditedPatient(patientToEdit, updatedMedicines);
+
+        model.setPatient(patientToEdit, editedPatient);
+        model.updateFilteredPatientList(PREDICATE_SHOW_ALL_PATIENTS);
+
+        return new CommandResult(generateSuccessMessage(editedPatient));
```
Comment (Bryce-3D) [inline src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:119]: Hi! Would it be faster if the code instead checked if `medicineToRemove` directly via `currentMedicines.contains(medicineToRemove)`? Though I suppose that this might have been done since it is easier to implement and runtime is not a concern in CS2103T?

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java (hunk 3 above)
Comment (Bryce-3D) [reply src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:37]: What are your thoughts on making the "unprescribe all" functionality be tied to not having any medicine argument passed?

File: src/main/java/seedu/address/model/medicine/Medicine.java [hunk 5]
```diff
@@ -59,4 +59,11 @@ public String toString() {
         return '[' + medName + ']';
     }
 
+    /**
+     * Get medication name.
+     */
+    public String getMedicineName() {
+        return medName;
+    }
```
Comment (Bryce-3D) [inline src/main/java/seedu/address/model/medicine/Medicine.java:67]: Hi! Should this be moved up in the file since it is a getter?

**Review by Bryce-3D — COMMENTED** [review]
Hi~

Generally looks good to me!

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java [hunk 6]
```diff
@@ -11,28 +15,42 @@
 import seedu.address.model.Model;
 import seedu.address.model.medicine.Medicine;
 import seedu.address.model.patient.Patient;
+
 /**
- * Removes all medication from an existing patient in the address book.
+ * Removes a medication from an existing patient in the address book.
```
Comment (Luoq1-Xu) [inline src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:20]: Perhaps would be better to phrase as "Removes the specified medication from an existing patient in the address book"

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java [hunk 7]
```diff
@@ -44,30 +62,93 @@ public CommandResult execute(Model model) throws CommandException {
...
+
+        for (Medicine medicine : currentMedicines) {
+            if (medicine.equals(medicineToRemove)) {
+                medicineFound = true;
+            } else {
+                updatedMedicines.add(medicine);
+            }
+        }
+
+        if (!medicineFound) {
```
Comment (Luoq1-Xu) [inline src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:109]: Do remember to add tests for this case. Also, we will likely move towards displaying only the patient's name (instead of the whole patient string representation), so might need to change this (and other parts) later

File: src/test/java/seedu/address/logic/commands/UnprescribeCommandTest.java [hunk 8]
```diff
@@ -43,59 +43,108 @@ public void execute_patientWithMedicines_success() {
...
+    }
+
+    @Test
+    public void execute_removingAllMedicines_success() {
+        // First, add multiple medicines to the patient
+        Patient patientToUpdate = model.getFilteredPatientList().get(INDEX_FIRST_PATIENT.getZeroBased());
+
+        // Create a set with multiple medicines and update the patient
+        HashSet<Medicine> medicineSet = new HashSet<>();
+        medicineSet.add(medicine);
```
Comment (Luoq1-Xu) [inline src/test/java/seedu/address/logic/commands/UnprescribeCommandTest.java:66]: Sorry if I missed it but where does the `medicine` variable come from?

File: src/test/java/seedu/address/logic/commands/UnprescribeCommandTest.java [hunk 9]
```diff
@@ -43,59 +43,108 @@ public void execute_patientWithMedicines_success() {
...
+        medicineSet.add(new Medicine("Aspirin"));
+        Patient patientWithMedicines = new PatientBuilder(patientToUpdate).withMeds(medicineSet).build();
+        model.setPatient(patientToUpdate, patientWithMedicines);
+
+        // Now test unprescribing all medicines
+        UnprescribeCommand unprescribeCommand = new UnprescribeCommand(INDEX_FIRST_PATIENT,
+                new Medicine("all"));
```
Comment (Luoq1-Xu) [inline src/test/java/seedu/address/logic/commands/UnprescribeCommandTest.java:73]: Having to initialise a medicine named "all" to use the `unprescribe all` function seems a little clunky. Perhaps we can discuss again and see if we can come up with a better approach (something like `Medicine.ALL`?)

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java [hunk 10]
```diff
@@ -44,30 +62,93 @@ public CommandResult execute(Model model) throws CommandException {
...
 
-        return new CommandResult(generateSuccessMessage(editedPatient, isEmpty));
+        return new CommandResult(generateSuccessMessageForAll(editedPatient));
+    }
+
+    /**
+     * Removes a specific medication from the patient
+     */
+    private CommandResult removeSpecificMedication(Model model, Patient patientToEdit) throws CommandException {
+        Set<Medicine> currentMedicines = patientToEdit.getMedicines();
```
Comment (Eun777) [inline src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:97]: Would it be good to check if the medicine is null using assertEquals before continuing?

File: src/test/java/seedu/address/logic/commands/UnprescribeCommandTest.java (hunk 8 above)
Comment (Eun777) [reply src/test/java/seedu/address/logic/commands/UnprescribeCommandTest.java:66]: I think `medicine` variable is from line 32 of this file.

File: src/test/java/seedu/address/logic/parser/UnprescribeCommandParserTest.java [hunk 11]
```diff
@@ -19,22 +20,57 @@ public class UnprescribeCommandParserTest {
...
+        Index targetIndex = INDEX_FIRST_PATIENT;
+        String userInput = targetIndex.getOneBased() + " " + PREFIX_MEDICINE;
+        String expectedMessage = "Medicine names should be alphanumeric";
+
+        assertParseFailure(parser, userInput,
+                expectedMessage);
+    }
+
+    @Test
+    public void parse_missingMedicinePrefix_throwsParseException() {
```
Comment (Eun777) [inline src/test/java/seedu/address/logic/parser/UnprescribeCommandParserTest.java:49]: Maybe an additional test where both the prefix and medicine name are missing can be added as an additional edge case test? Not sure if it has already been added but a separate test could be beneficial.

**Review by Eun777 — CHANGES_REQUESTED** [review]
Generally changes look good, some additional tests and checks could be beneficial for edge cases!

File: src/test/java/seedu/address/logic/commands/UnprescribeCommandTest.java (hunk 8 above)
Comment (Luoq1-Xu) [reply src/test/java/seedu/address/logic/commands/UnprescribeCommandTest.java:66]: haha `blindness +5`
Thanks

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java (hunk 2 above)
Comment (szelingtan) [reply src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:70]: okie!

File: src/main/java/seedu/address/logic/commands/UnprescribeCommand.java (hunk 4 above)
Comment (szelingtan) [reply src/main/java/seedu/address/logic/commands/UnprescribeCommand.java:119]: okie!
